Deprecate k_mem_pool API, remove sys_mem_pool allocator

andyross opened the pull request:
The k_heap/sys_heap code has been default for a release, so it's time to deprecate the old allocator.

This eliminates the older CONFIG_MEM_POOL_HEAP_BACKEND kconfig, effectively making it true always and removing the ability for applications to request the older pool engine.  It then deprecates the older APIs, leaving them in place for internal users as a "Z" API.

There are still a few users of the older API internally that will need cleanup (**edit: not anymore, it's 100% gone in the current version**).  I'll submit those separately I think, I'm not sure if those are in scope for 2.4 or not:

* k_malloc and the system pool are still implemented in terms of mem_pool, even though those are just wrappers now.  It should be directly on top of a k_heap.

* The per-thread resource pools likewise are mem_pools still.  This should be k_heap's instead.

* Mailbox has a API that will retrieve a message into a newly-allocated heap block.  This API is weird and itself deprecated, but if we decide we actually want to preserve the functionality it will need some kind of new variant.

* Similarly pipe has block usage internally.

And there are two places where the underlying sys_heap allocator still gets used:

* The minimal libc malloc() implementation is on top of sys_heap instead of k_malloc.  I'm not sure why, but this needs some kind of port.

* The lib/gui/lvgl code has an internal sys_mem_pool and needs a port to sys_heap (not hard to write, but I have no idea how to test it)

Fixes #24358

Commit:
kernel: Make the "heap" backend to mem_pool default

Remove the MEM_POOL_HEAP_BACKEND kconfig, treating it as true always.
Now the legacy mem_pool cannot be enabled and all usage uses the
k_heap/sys_heap backend.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel: Remove mem_pool tests

This API is being deprecated, and the underlying sys_heap code has its
tests elsewhere.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Deprecate k_mem_pool APIs

Mark all k_mem_pool APIs deprecated for future code.  Remaining
internal usage now uses equivalent "z_mem_pool" symbols instead.

Fixes #24358

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Ugh, it collides.  Sorry, I did this on a week-stale master.  Will get that cleaned up.

Commit:
lib/gui/lvgl: Use a sys_heap and not a sys_mem_pool

The sys_mem_pool allocator is a legacy thing.  Use the standard heap
to reduce code size.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

pabigot commented:
I'm not clear on how #24358 is a bug, let alone having a release-blocking priority, except as a means of getting this late change into 2.4.0.  And I've been told before that stage of release cycle is not a valid reason for rejecting a PR.

But from policy discussion heard at the last TSC regarding criteria for accepting late addition of new boards I need to ask: If a PR like this came three days before release from anybody other than a platinum member company, would people be seriously considering merging it into the release when it passes?

andrewboie commented:
> I'm not clear on how #24358 is a bug

It's a bug because we have two mechanisms in the kernel which do the same thing, and we do not want to maintain both, particularly not in the second LTS release.

nashif commented:
> If a PR like this came three days before release from anybody other than a platinum member company, would people be seriously considering merging it into the release when it passes?

seriously? What exactly is the purpose of creating releases and having almost 4 weeks of bug fix and stabilization period if we do not address issues like this? 

This was a bug when it was created in April, move to an Enhancement because it did not happen for 2.3 with clear documentation in the issue itself that it should go into 2.4. This had nothing to do with membership level and it is disappointing to hear that TBH. This is not something someone is pushing for to get a feature in late in the game, this actually removes duplication and confusion. So there is no direct and "personal" gain from doing this.

pabigot reviewed (commented):
Looking closely it seems the impact would only be on applications that explicitly selected `CONFIG_MEM_POOL_HEAP_BACKEND=n`, so once the missed in-tree uses are fixed and it passes I don't see any technical reason it shouldn't go in.

pabigot commented:
Seriously, the need to do this has been known for months.  Why is it being done so late, and why is it being forced/enabled by giving #24358 a release-blocking priority?  Zephyr's ability to plan what goes into a release, and track its progress, is disappointingly weak.

I do feel that part of the criteria for accepting PRs in the stabilization period should be the risk of destabilizing, invalidating all the testing that's been done, and introducing build warnings in out of tree applications and downstream forks that have been faithfully tracking the release and suddenly have to make changes to accommodate a last minute API deprecation.

That there are so many remaining in-tree uses that fail CI because of the new deprecation suggests this isn't a baseless concern.

And I remain very skeptical that this would be getting this level of attention and support if a random community contributor had proposed it at this point.

tyrantlizard77 reviewed (changes requested):
Just noticed: this needs to deprecate all the `sys_mem_pool` APIs as well as they are superseded by `sys_heap`.

If you want to remove the `sys_mem_pool` stuff in another PR please remove the "Fixes" line in the PR description. I'd rather it was removed here though.

This PR was discussed in the TSC today, the upshot is that we will wait until the merge window opens for 2.5 before merging. And then completely remove for 2.6, with a documented exception to the two release rule. The 2.4 release notes will indicate these APIs are now considered deprecated even though they will not be tagged with `__deprecated` until 2.5.

andrewboie commented:
This also needs to deprecate `k_pipe_block_put()`

AFAICT the full set of APIs to deprecate is:
  * `k_mbox_data_block_get()`
  * `k_pipe_block_put()`
  * `K_MEM_POOL_DEFINE()`
  * `k_mem_pool_alloc()`
  * `k_mem_pool_free()`
  * `k_mem_pool_free_id()`
  * `SYS_MEM_POOL_DEFINE()`
  * `sys_mem_pool_init()`
  * `sys_mem_pool_alloc()`
  * `sys_mem_pool_free()`
  * `sys_mem_pool_try_expand_inplace()`

Since this is now in scope for early 2.5 and not 2.4, I would prefer if these deprecation PRs converted internal kernel usage to k_heap / sys_heap rather than introducing these z_ variants.

Commit:
kernel/k_malloc: Decouple k_malloc/k_free from mem_pool

These were implemented in terms of the mem_pool/block API directly
(for complicated reasons, the pointers returned from this API may have
been allocated from allocators other than the single system heap).
Have them use a k_heap instead.

Requires a tweak to one test which had hard-coded an assumption about
the header size.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

carlescufi commented:
@andyross @andrewboie could we pick this up now that 2.4 has been released? let's not wait until the next release is close yet again.

Commit:
kernel: Make thread resource pools into sys_heaps

The k_mem_pool allocator is no more, and the z_mem_pool compatibility
API is going away.  The internal allocator should be a k_heap always.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

MaureenHelm commented:
> @andyross @andrewboie could we pick this up now that 2.4 has been released? let's not wait until the next release is close yet again.

Let's get this moving again

nashif commented:
> Let's get this moving again

@andyross  already working on this.

Commit:
tests/lib/heap: Add test for sys_heap_realloc()

Simple unit test for the realloc functionality, covering all state
transitions.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/mem_protect: Remove mem_pool test case

The sys_mem_pool data structure is going away.  And this test case
didn't actually do much.  All it did was create a sys_mem_pool in the
app data section (I guess that's the "mem_protect" part?) and validate
that it was usable.  We have tests for sys_heap to do that already
elsewhere anyway; no point in porting.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel: Remove sys_mem_pool test

This data structure is going away, and its replacement (sys_heap) has
tests already.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
samples/userspace/prod_consumer: Use a sys_heap

This code used a sys_mem_pool directly.  Use a new-style heap instead
to do the same thing.

(Note that the usage is a little specious -- it allocates from the
heap but doesn't appear to fill or check any data therein, just that
the heap memory can be copied from the two memory domains.  It's
unclear exactly what this is trying to demonstrate and we might want
to improve the sample to do something less trivial.)

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
lib: Remove sys_mem_pool implementation

This has been replaced by sys_heap now and all dependencies are gone.
Remove.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/mem_protect: Port to k_heap API

The kernel resource pool is now a k_heap.  There is a compatibility
API still, but this is a core test that should be exercising the core
API.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/mheap_api_concept: Remove whitebox tests

These two test cases were making whitebox assumptions of both the
block header size and memory layout of an old-style k_mem_pool that
aren't honored by the k_heap allocator.  They aren't testing anything
that isn't covered elsewhere.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/thread_stack: Increase heap size

This test was written to use a TINY system heap (64 bytes) from which
it has to allocate on behalf of a userspace process.  The change in
convention from mem_pool (where the byte count now includes metadata
overhead) means it runs out of space.  Bump to 192 bytes.  Still tiny.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/net/socket/udp: Increase kernel heap size

On userspace platforms, this test needs a little bit of kernel heap.
The old mem_pool number was specified without metadata overhead
(i.e. it reflected 128 bytes of actual data available and the metadata
was stored silently somewhere else), where the new heap specifies the
size of the contiguous buffer in memory that stores both data and
chunk headers, etc...

Increase to 256 bytes.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
lib/os/heap: Add sys_heap_realloc()

Add an optimized realloc() implementation that can successfully expand
allocations in place if there exists enough free memory after the
supplied block.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
libc/minimal: Use a sys_heap for the malloc implementation

The older sys_mem_pool is going away and being replaced by a new
allocator.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Resurrect this guy with a more invasive change.  Contains the old deprecation, and also a complete removal of the older sys_mem_pool allocator backend.  That code (and a bunch of tests specific to it) is now gone, the only allocator we ship is the new one.

The compatibility API (which implements the old scheme in terms of the new allocator) remains in place as "z_mem_pool" for the benefit of a few in-tree spots that still use it and are difficult for me to test directly (mostly driver code for devices I don't have).  We can address these piecewise as we find maintainers able to validate, that stuff is all boilerplate.

And this does not contain a patch that actually removes the deprecated APIs.  I'm still not clear on the project management intent there as to whether that should happen this release or next.

nashif commented:
```

/workdir/zephyr/tests/kernel/mem_protect/syscalls/src/main.c:392:2: error: 'z_thread_resource_pool_assign' is deprecated [-Werror=deprecated-declarations]
--



```

andyross commented:
> did you miss one?

Yeah, last minute goof.  That function is an internal wrapper, it's not supposed to be deprecated.  This happened because there had been a "temporarily undeprecated" k_* function there at one point (i.e. to suppress the deprecation warnings while I worked on somethign else), which worked by commenting the function out with a C++ comment I knew would get caught later (e.g. "//__deprecated // FIXME reenable"), which did indeed get caught in a final checkpatch.  But then I blindly reenabled it forgetting that this was now an internal function that didn't need to be deprecated and pushed before re-running the whole sanitycheck.

Probably too much detail.  Easy fix.  Will push as soon as my rebased tree completes its check.

Commit:
subsys/bluetooth: Add missing stdint.h include

This tiny header uses non-builtin types but includes no headers that
would define them.  Recent header motion seems to have exposed a case
where this file can get built before its dependencies are included.
Add the header directly.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
I'd been assuming those weird BSIM errors (in code that has nothing to do with mem_pool) were an upstream thing, but they turned out to be a latent bug with a header (missing stdint.h) that the removal of mempool_sys.h exposed.  Fixed.

andrewboie commented on include/kernel.h:
```diff
@@ -607,13 +607,19 @@ extern FUNC_NORETURN void k_thread_user_mode_enter(k_thread_entry_t entry,
+	thread->resource_pool = heap;
+}
+
+static inline void z_thread_resource_pool_assign(struct k_thread *thread,
```
this will break existing users.
better I think to leave this function name as `k_thread_resource_pool_assign` and mark as `__deprecated`

tyrantlizard77 reviewed (approved):
one nit about an API name change. rest looks fine.

I think we will need to file bugs or otherwise figure out what to do about the remaining uses of z_mem_pool_* inside the kernel so that we can remove all the z_mem_pool code entirely for LTS2, but we have time.

nashif commented on drivers/usb/device/usb_dc_kinetis.c:
```diff
@@ -80,7 +80,7 @@ static struct buf_descriptor __aligned(512) bdt[(NUM_OF_EP_MAX) * 2 * 2];
 #define EP_BUF_NUMOF_BLOCKS		(NUM_OF_EP_MAX / 2)
 
-K_MEM_POOL_DEFINE(ep_buf_pool, 16, 512, EP_BUF_NUMOF_BLOCKS, 4);
+Z_MEM_POOL_DEFINE(ep_buf_pool, 16, 512, EP_BUF_NUMOF_BLOCKS, 4);
```
shouldn we be using K_HEAP_DEFINE here?

andyross replied:
Yes, but I don't feel comfortable doing the port without a device with this USB controller to validate.  The wrappers are near-zero-size and very low risk.  Changing API usage blindly seems like more downside than up.

jfischer-phytec-iot replied:
I did https://github.com/jfischer-phytec-iot/zephyr/commit/7f482950a000072d2b3ba97b080d9b8de0fed358 just for fun. IMHO K_HEAP_DEFINE is not an easy replacement for K_MEM_POOL_DEFINE as the overhead for the heap management has to be considered. However we want to remove this intermediate buffers anyway, so I am fine to do minimal changes and use Z_MEM_POOL_DEFINE for now.

andyross replied:
There's a port of that subsystem (and all the others) in this patch now.  But yes: in the old style you would specify how many objects you wanted to be able to allocate, and the overhead would be generated magically.  Now you specify how many bytes you want to use for heap storage and the overhead is included in that.

nashif commented on drivers/usb/device/usb_dc_kinetis.c:
```diff
@@ -353,14 +353,14 @@ int usb_dc_ep_configure(const struct usb_dc_ep_cfg_data * const cfg)
 
 	if (bdt[idx_even].buf_addr) {
-		k_mem_pool_free(block);
+		z_mem_pool_free(block);
```
why are we using this internal API and not k_heap_free?

cvinayak commented on subsys/bluetooth/controller/hal/cntr.h:
```diff
@@ -4,6 +4,7 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
+#include <stdint.h>
```
```suggestion
```

Include `#include <stdint.h>` in c files where needed. This is not right to have here to satisfy files that include `#include "hal/cntr.h"`

That said, I will have a future action item on myself to clean up other redundant/incorrect includes in controller hal folder, like sys/dlist.h etc  which is never used in cntr.c file.

andyross replied:
Missed this comment.  There is a uint8_t and a uint32_t usage six lines down in this same file (the arguments to cntr_cmp_set()).  There is not way for this to compile correctly in all cases without this include here.

Commit:
kernel: Remove legacy mem_pool usage

The mailbox and msgq utilities had API variants that could pass old
mem_pool blocks through the data structure.  That API is being
deprected (and the features were obscure), so remove the internal
support.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: subsys: lib: drivers: Use k_heap instead of z_mem_pool wrappers

Use the core k_heap API pervasively within our tree instead of the
z_mem_pool wrapper that provided compatibility with the older mempool
implementation.

Almost all of this is straightforward swapping of one alloc/free call
for another.  In a few cases where code was holding onto an old-style
"mem_block" a local compatibility struct with a single field has been
swapped in to keep the invasiveness of the changes down.

Note that not all the relevant changes in this patch have in-tree test
coverage, though I validated that it all builds.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/pipe/pipe_api: Remove mem_pool-related test cases

Remove test cases that exercise the deprecated mem_pool features of
the pipe utility.

Note that this leaves comparatively few cases left, we should probably
audit coverage after this merges and rewrite tests that aren't
interdependent.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Remove k_mem_pool API

Remove the declarations for the older mempool API

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Remove z_mem_pool wrapper internals

These implemented a k_mem_pool in terms of the now universal k_heap
utility.  That's no longer necessary now that the k_mem_pool API has
been removed.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Resurrect this guy with a rebase, and a few extra patches that port the tree internals away from the old API, remove the old API entirely, and then remove the wrappers.  Should be good to go per existing comments.  We'll see what CI says.

andyross commented:
Heh, usb_dc_nrfx was one of the blind changes, isn't part of a default sanitycheck, but is part of CI.  Fixed typo, repushed.

nashif commented:
nrf52_bsim is unrelated and caused by #29810

andyross commented:
Next spin: now-unused variable warning in hid-cdc sample
